rfcs/2021-10-29-9465-splunk-hec-indexer-acknowledgements.md

```diff
@@ -0,0 +1,346 @@
+    intend to move data from a `splunk_hec` source to a `splunk_hec` sink,
+    passing through the channel IDs can be helpful.
+  * Rather than assign `ackId`’s per channel, the `splunk_hec` source will
+    assign `ackId`’s across all its requests. In other words, if the source
+    receives two requests with differing channel IDs, it will reply with `ackId
+    = 0` and `ackId = 1` rather than `ackId = 0` and `ackId = 0`.
+
+* Pending acks
+  * We will support the overall `max_number_of_acked_requests_pending_query`
+    configuration described in Splunk documentation (`max_pending_acks`) to
+    avoid memory issues. We will not support the channel-based settings.
+
+### Implementation
+
+Implementation details mostly concern creating, storing, and updating `ackId`’s
+and statuses. The following suggested implementation prioritizes memory
+efficiency.
+
+#### Data Structures
+
+* A [bitvec](https://docs.rs/bitvec/0.22.3/bitvec/) (or similar structure)
+  `ack_ids_in_use` whose indices represent `ackId`’s, max size is determined by
+  `max_pending_acks`, and is initialized with all `0`'s'. A value of `0` at an
+  index indicates that the `ackId` is not in-use. A value of `1` indicates
+  in-use.
+* A bitvec `ack_ids_ack_status` whose indices also correspond to `ackId`’s, max
+  size is determined by `max_pending_acks`, and is initialized with all `0`'s. A
+  value of `0` at an index indicates that the associated request data has not
+  yet been delivered. A value of `1` indicates that the data has been delivered.
+  These two bitvecs will be wrapped in a struct with useful methods. A single
+  bitvec is not necessarily sufficient as we care about 3 states: `ackId` is
+  available, `ackId` is pending/dropped, and `ackId` is delivered.
+* An index pointer `currently_available_ack_id` initialized with `0`.
+
+#### Process
+
+* Assigning and updating `ackId`’s and statuses
+  * To assign a new `ackId` to a new request, we use the
+    `currently_available_ack_id` current value and set
+    `ack_ids_in_use[currently_available_ack_id] = 1`. We increment
+    `currently_available_ack_id` to the next index. We respond to the request
+    with a `200 OK` with the `ackId` included in the JSON body.
+    * If incrementing causes `currently_available_ack_id` to exceed the size of
+      the bitvec, we wrap around to index `0` and search for the nearest
+      available `ackId` (i.e. where `ack_ids_in_use[index] = 0`).
+    * If there are no available `ackId`’s, we can begin to drop pending, in-use
+      `ackId`’s, setting `ack_ids_in_use[...] = 0`. For simplicity, we can drop
+      starting from the lowest `ackId` values (though there may be edge cases
+      here that can cause this to continuously drop the same `ackId`'s). We
+      should drop a chunk of pending `ackId`’s rather than a single `ackId` at a
+      time to avoid potentially re-scanning the `ack_ids_in_use` array on the
+      next incoming request.
```

Comment:
It seems like maybe we should start dropping at `currently_available_ack_id + 1`? I think that would, generally, result in the oldest request ack being dropped, which would be the most desirable one to drop.

Reply:
I see, `currently_available_ack_id + 1` is probably a better proxy for oldest; I like the idea. There may be some corner case issues wherein `currently_available_ack_id` wraps around and settles in the middle of `ack_ids_in_use`. Then, if dropping is needed soon after, the beginning of `ack_ids_in_use` would be oldest but would not be dropped.

Reply:
Updated in 4c6ec08, thanks!